### PR TITLE
Add environment parse function that supports default value

### PR DESCRIPTION
**Pull request by kwen2501**
We use "-2" to represent an unset environment variable.
Now adding a util function to attach default value if environment variable is unset.

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -624,7 +624,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
       "ProcessGroupNCCL is only supported with GPUs, no GPUs found!");
   blockingWait_ = parseEnvVarFlag(NCCL_BLOCKING_WAIT);
   asyncErrorHandling_ =
-      static_cast<ErrorHandlingMode>(parseEnvVarInt(NCCL_ASYNC_ERROR_HANDLING));
+      static_cast<ErrorHandlingMode>(parseEnvVarIntDefault(NCCL_ASYNC_ERROR_HANDLING, 0));
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:627]: Just curious, right not when it is not set, where does it default to value -2?

**Review by H-Huang — APPROVED** [review]
LGMT thanks!

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:627]: Here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/Utils.hpp#L88

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Fail test_DistributedDataParallel (main.TestDistBackendWithSpawn)' -c landrace

I suspect that this is a land race because the test fails in pull request itself, `merge -g` only hide the failure until it reaches trunk.  From what I remember, this test used to be flaky https://github.com/pytorch/pytorch/issues/77317, so there might be something fishy uncovered by this change, not the change itself:

* https://hud.pytorch.org/pytorch/pytorch/commit/784f4ba1ce16996d497fae2fb107425b3bbeb71b
* https://hud.pytorch.org/pytorch/pytorch/commit/755b39ba6668d2ce8c0e3ba4ee032ff0dfc827b7
* https://hud.pytorch.org/pytorch/pytorch/commit/b8d2ab3dd5869e0af3aa9490636acf887d664be7
* https://hud.pytorch.org/pytorch/pytorch/commit/a0be0ca16144fababbbb94f58b1fd88e41f29162
* https://hud.pytorch.org/pytorch/pytorch/commit/18d8c548f4458c03ae6be60903db3ccf1c0d8d3f

As the first step, please rebase this one

**Comment by kwen2501** [comment]
@huydhn The previous land failure indeed seems to be caused by flakiness as the CI is all green now. (I just reran it without any change.)

Can I re-merge now?

**Pull request closed without merging**